src/backoffice/shared/components/tree/tree.context.ts

```diff
@@ -23,15 +25,44 @@ export class UmbTreeContextBase implements UmbTreeContext {
 	#selection = new DeepState(<Array<string>>[]);
 	public readonly selection = this.#selection.asObservable();
 
-	repository!: UmbTreeRepository;
+	repository?: UmbTreeRepository;
+
+	#initResolver?: () => void;
+	#initialized = false;
+
+	#init = new Promise<void>((resolve) => {
+		this.#initialized ? resolve() : (this.#initResolver = resolve);
+	});
 
 	constructor(host: UmbControllerHostInterface, tree: ManifestTree) {
-		this.#host = host;
+		this.host = host;
 		this.tree = tree;
 
-		if (this.tree.meta.repository) {
-			// TODO: should be using the right extension and the createExtensionClass method.
-			this.repository = new this.tree.meta.repository(this.#host) as any;
+		const repositoryAlias = this.tree.meta.repositoryAlias;
+		if (!repositoryAlias) throw new Error('Tree must have a repository alias.');
+
+		new UmbObserverController(
+			this.host,
+			umbExtensionsRegistry.getByTypeAndAlias('repository', this.tree.meta.repositoryAlias),
+			async (repositoryManifest) => {
+				if (!repositoryManifest) return;
+
+				try {
+					const result = await createExtensionClass<UmbTreeRepository>(repositoryManifest, [this.host]);
+					this.repository = result;
+					this.#checkIfInitialized();
```

Comment:
I think we could make a AsPromise on the Observer Controller, which could be 'extended' by the given method returning a PromiseAlikeResult which would postpone its completion.

Then this promise could be used instead of the `#checkIfInitialized`.

Don't know if that concept can do us any good? but that was my thoughts on this one.

Reply:
I'll just create a task for it in Azure Boards. Then we can revisit in near future.